src/concurrency/concurrency.md

```diff
@@ -1,3 +1,537 @@
+        });
+
+        if state && !last_state {
+            last_state = state;
+
+            // Set PA1 high if we've seen a rising edge on PA0.
+            interrupt::free(|cs| {
+                let gpioa = MY_GPIO.borrow(cs).borrow();
+                gpioa.as_ref().unwrap().odr.modify(|_, w| w.odr1().set_bit());
+            });
+        }
+    }
+}
+
+#[interrupt]
+fn timer() {
+    // This time in the interrupt we'll just clear PA0.
+    interrupt::free(|cs| {
+        // We can use `unwrap()` because we know the interrupt wasn't enabled
+        // until after MY_GPIO was set; otherwise we should handle the potential
+        // for a None value.
+        let gpioa = MY_GPIO.borrow(cs).borrow();
+        gpioa.as_ref().unwrap().odr.modify(|_, w| w.odr1().clear_bit());
+    });
+}
+```
+
+That's quite a lot to take in, so let's break down the important lines.
+
+```rust
+static MY_GPIO: Mutex<RefCell<Option<stm32f405::GPIOA>>> =
+    Mutex::new(RefCell::new(None));
+```
+
+Our shared variable is now a `Mutex` around a `RefCell` which contains an
+`Option`. The `Mutex` ensures we only have access during a critical section,
+and therefore makes the variable Sync, even though a plain `RefCell` would not
+be Sync. The `RefCell` gives us interior mutability with references, which
+we'll need to use our `GPIOA`. The `Option` lets us initialise this variable
+to something empty, and only later actually move the variable in. We cannot
+access the peripheral singleton statically, only at runtime, so this is
+required.
+
+```rust
+interrupt::free(|cs| MY_GPIO.borrow(cs).replace(Some(dp.GPIOA)));
+```
+
+Inside a critical section we can call `borrow()` on the mutex, which gives us
+a reference to the `RefCell`. We then call `replace()` to move our new value
+into the `RefCell`.
+
+```rust
+interrupt::free(|cs| {
+    let gpioa = MY_GPIO.borrow(cs).borrow();
+    gpioa.as_ref().unwrap().odr.modify(|_, w| w.odr1().set_bit());
+});
+```
+
+Finally we use `MY_GPIO` in a safe and concurrent fashion. The critical section
+prevents the interrupt firing as usual, and lets us borrow the mutex.  The
+`RefCell` then gives us an `&Option<GPIOA>`, and tracks how long it remains
+borrowed - once that reference goes out of scope, the `RefCell` will be updated
+to indicate it is no longer borrowed.
+
+Since we can't move the `GPIOA` out of the `&Option`, we need to convert it to
+an `&Option<&GPIOA>` with `as_ref()`, which we can finally `unwrap()` to obtain
+the `&GPIOA` which lets us modify the peripheral.
+
+Whew! This is safe, but it is also a little unwieldy. Is there anything else
+we can do?
+
+## RTFM
+
+One alternative is the [RTFM framework], short for Real Time For the Masses. It
+uses interrupt priorities to statically ensure that shared resources are always
+accessed safely, without requiring the overhead of always entering critical
+sections and using reference counting (as in `RefCell`). This has a number of
```

Comment:
No reference "counting" here.

Reply:
I think it's fair to call what `RefCell` does "reference counting": it literally counts how many `&` and `&mut` have been taken to dynamically enforce the borrow rules. It's not the same as a reference counted pointer (like `Rc`), which counts references to know when to drop a value from the heap, but it's still reference counting.

Reply:
After reading the code I agree with your assessment about the implementation details which where not clear to me before. However I still think it is confusing to mention "reference counting" as an implementation detail here and would prefer to use a different phrase such as "maintaining the borrow state" to refer to the overhead incurred by using `RefCell`.